sdk/storage/Azure.Storage.Common.DataMovement/src/FilesystemScanner.cs

```diff
@@ -0,0 +1,116 @@
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Text;
+
+namespace Azure.Storage.Common.DataMovement
+{
+    /// <summary>
+    /// FilesystemScanner class
+    /// </summary>
+    public static class FilesystemScanner
+    {
+        /// <summary>
+        /// Enumerates all files pointed to by the provided path, including those in subdirectories (if path is a directory).
+        /// </summary>
+        /// <param name="path">Filesystem location.</param>
+        /// <returns>Enumerable list of absolute paths containing all relevant files the user has permission to access.</returns>
+        public static IEnumerable<string> ScanLocation(string path)
+        {
+            // Path type is ambiguous at start
+            bool isDirectory = false;
+
+            try
+            {
+                // Make sure we're dealing with absolute path
+                path = Path.GetFullPath(path);
+
+                // Check if path points to a directory
+                if ((File.GetAttributes(path) & FileAttributes.Directory) == FileAttributes.Directory)
+                {
+                    isDirectory = true;
+                }
+            }
+            catch
+            {
+                // If there's an error here, there aren't any valid entries to scan at the
+                // given path; either the path is invalid/nonexistant, or it isn't accessible by the user. In this case,
+                // don't return anything.
+                yield break;
+            }
+
+            // If we're given a directory, parse its children recursively
+            if (isDirectory)
+            {
+                // Create a queue of folders to enumerate files from, starting with provided path
+                Queue<string> folders = new Queue<string>();
+                folders.Enqueue(path);
+
+                while (folders.Count > 0)
+                {
+                    // Grab a folder from the queue
+                    string dir = folders.Dequeue();
+
+                    // Try to enumerate and queue all subdirectories of the current folder
+                    try
+                    {
+                        foreach (string subdir in Directory.EnumerateDirectories(dir))
+                        {
+                            folders.Enqueue(subdir);
+                        }
+                    }
+                    // If we lack permissions to enumerate, skip this folder
+                    catch
+                    {
+                        continue;
+                    }
+
+                    foreach (string file in Directory.EnumerateFiles(dir))
+                    {
+                        // Try to open a stream to the file to test read permissions
+                        try
+                        {
+                            using (FileStream fs = File.Open(file, FileMode.Open, FileAccess.Read))
+                            {
+                                goto readSuccess;
```

Comment:
Is the reason why you're doing this is to keep the stream open? Maybe we should do this check later when we actually have to open the file because Murphy's Law (people can go and change the access control on the file before we go to upload even though we checked here).

Reply:
I thought that the stream automatically closes once that using block is exited, is that wrong? I did it because I was using File.Exists(path) before since it claimed to be able to discern read permissions in the documentation, but that didn't hold up when I was testing (it just told us whether or not the file literally existed) and I needed another approach.

So, it seemed like one option was looking through ACL for the file, but then I saw a file could deny reading ACL while allowing reading contents. Instead, I opted to just attempt to open a stream with read permissions, as it should use that check to know whether or not to add it to the enumeration for now, and then close the stream.

If it isn't possible to read later, I was assuming that could just be handled with log + skip per user options, as I wasn't sure if there was any better alternative to dealing with file changes from scan to operation time.

Reply:
I should have clarified my question. In the body of this stream you have it jump to `readSuccess` which is fine, but is there a reason why you did this over just returning the file in the body of this, or even just checking if it just opens and then closing it right after.
e.g. something like this
```
try
{
    File.Open(this.DataSourceFileName, FileMode.Open, FileAccess.Read).Dispose();
    return true;
}
```
Also is there a reason why we picked `Open` over `OpenRead`?

If it's not possible to read later, then this check is unnecessary right? We only care if we can access the file when we actually want to read it.

Reply:
Nope, no reason for either of those. I'm changing the scanner to function as you wrote; I think I was overthinking something about the try-catch and yield return going together.

I suppose so; I was thinking it would save time for the operations later, but really it wouldn't be that intensive at all. So, the only files the scanner really wouldn't be able to return are those inside of a subfolder that we can't read, since we wouldn't be able to get inside to enumerate. However, if a file exists somewhere we are able to at least see that it exists, we want to put it in the list and leave it to the method caller to decide whether reads will or won't happen. Does that sound like the desired functionality? I'll drop the read checks here if that's the case.

Reply:
I took the file checking code out of the scanner for now, to avoid any mismatches between scan time and operation time as you mentioned. The functionality now is to simply return all files we have permission to see exists, and let the caller handle the question of whether or not we can work on the file.